Building null origins in RelyingParty

I think you should be able to build the [RelyingParty](https://github.com/Yubico/java-webauthn-server/blob/main/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingParty.java) object using a null Origins Set (it just takes the RP ID if it is). However, the builder will not allow it `Caused by: java.lang.NullPointerException: origins is marked non-null but is null`. I guess that is because of the nonnull annotation on the origins field.

Version 2.5.0.

--- a/NEWS
+++ b/NEWS
@@ -1,5 +1,6 @@
 == Version 2.7.0 (unreleased) ==
 
+* Added overloaded setter `RelyingPartyBuilder.origins(Optional<Set<String>>)`.
 * (Experimental) Added a new suite of interfaces, starting with
   `CredentialRepositoryV2`. `RelyingParty` can now be configured with a
   `CredentialRepositoryV2` instance instead of a `CredentialRepository`

--- a/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingParty.java
+++ b/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingParty.java
@@ -588,6 +588,16 @@ public class RelyingParty {
         Optional.empty();
     private @NonNull Optional<AttestationTrustSource> attestationTrustSource = Optional.empty();
 
+    public RelyingPartyBuilder origins(@NonNull Set<String> origins) {
+      this.origins = origins;
+      return this;
+    }
+
+    public RelyingPartyBuilder origins(Optional<Set<String>> origins) {
+      this.origins = origins.orElse(null);
+      return this;
+    }
+
     public static class MandatoryStages {
       private final RelyingPartyBuilder builder = new RelyingPartyBuilder();
 

--- a/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingPartyV2.java
+++ b/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingPartyV2.java
@@ -563,6 +563,16 @@ public class RelyingPartyV2<C extends CredentialRecord> {
         Optional.empty();
     private @NonNull Optional<AttestationTrustSource> attestationTrustSource = Optional.empty();
 
+    public RelyingPartyV2Builder<C> origins(@NonNull Set<String> origins) {
+      this.origins = origins;
+      return this;
+    }
+
+    public RelyingPartyV2Builder<C> origins(Optional<Set<String>> origins) {
+      this.origins = origins.orElse(null);
+      return this;
+    }
+
     /**
      * The extension input to set for the <code>appid</code> and <code>appidExclude</code>
      * extensions.

--- a/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
+++ b/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
@@ -134,7 +134,7 @@ public class RelyingPartyTest {
   }
 
   @Test
-  public void testOriginsWithEmptyOptionalSet() {
+  public void testOriginsWithAbsentOptionalSet() {
     RelyingParty rp =
         RelyingParty.builder()
             .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())

--- a/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
+++ b/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
@@ -104,6 +104,64 @@ public class RelyingPartyTest {
   }
 
   @Test
+  public void testOriginsWithEmptySet() {
+    Set<String> origins = new HashSet<>();
+
+    RelyingParty rp =
+        RelyingParty.builder()
+            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+            .credentialRepository(unimplementedCredentialRepository())
+            .origins(origins)
+            .build();
+
+    assertEquals(0, rp.getOrigins().size());
+  }
+
+  @Test
+  public void testOriginsWithSet() {
+    Set<String> origins = new HashSet<>();
+    origins.add("test1");
+    origins.add("test2");
+
+    RelyingParty rp =
+        RelyingParty.builder()
+            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+            .credentialRepository(unimplementedCredentialRepository())
+            .origins(origins)
+            .build();
+
+    assertEquals(2, rp.getOrigins().size());
+  }
+
+  @Test
+  public void testOriginsWithEmptyOptionalSet() {
+    RelyingParty rp =
+        RelyingParty.builder()
+            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+            .credentialRepository(unimplementedCredentialRepository())
+            .origins(Optional.empty())
+            .build();
+
+    assertEquals(1, rp.getOrigins().size());
+  }
+
+  @Test
+  public void testOriginsWithOptionalSet() {
+    Set<String> origins = new HashSet<>();
+    origins.add("test1");
+    origins.add("test2");
+
+    RelyingParty rp =
+        RelyingParty.builder()
+            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+            .credentialRepository(unimplementedCredentialRepository())
+            .origins(Optional.of(origins))
+            .build();
+
+    assertEquals(2, rp.getOrigins().size());
+  }
+
+  @Test
   public void filtersAlgorithmsToThoseAvailable() throws HexException {
     for (Provider prov : Security.getProviders()) {
       if (prov.getName().contains("EC")) {

--- a/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
+++ b/webauthn-server-core/src/test/java/com/yubico/webauthn/RelyingPartyTest.java
@@ -135,14 +135,26 @@ public class RelyingPartyTest {
 
   @Test
   public void testOriginsWithAbsentOptionalSet() {
-    RelyingParty rp =
-        RelyingParty.builder()
-            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
-            .credentialRepository(unimplementedCredentialRepository())
-            .origins(Optional.empty())
-            .build();
+    {
+      RelyingParty rp =
+          RelyingParty.builder()
+              .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+              .credentialRepository(unimplementedCredentialRepository())
+              .origins(Optional.empty())
+              .build();
+      assertEquals(1, rp.getOrigins().size());
+    }
 
-    assertEquals(1, rp.getOrigins().size());
+    {
+      RelyingPartyV2<RegisteredCredential> rp =
+          RelyingParty.builder()
+              .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+              .credentialRepositoryV2(
+                  new CredentialRepositoryV1ToV2Adapter(unimplementedCredentialRepository()))
+              .origins(Optional.empty())
+              .build();
+      assertEquals(1, rp.getOrigins().size());
+    }
   }
 
   @Test
@@ -151,14 +163,26 @@ public class RelyingPartyTest {
     origins.add("test1");
     origins.add("test2");
 
-    RelyingParty rp =
-        RelyingParty.builder()
-            .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
-            .credentialRepository(unimplementedCredentialRepository())
-            .origins(Optional.of(origins))
-            .build();
+    {
+      RelyingParty rp =
+          RelyingParty.builder()
+              .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+              .credentialRepository(unimplementedCredentialRepository())
+              .origins(Optional.of(origins))
+              .build();
+      assertEquals(2, rp.getOrigins().size());
+    }
 
-    assertEquals(2, rp.getOrigins().size());
+    {
+      RelyingPartyV2<RegisteredCredential> rp =
+          RelyingParty.builder()
+              .identity(RelyingPartyIdentity.builder().id("localhost").name("Test").build())
+              .credentialRepositoryV2(
+                  new CredentialRepositoryV1ToV2Adapter(unimplementedCredentialRepository()))
+              .origins(Optional.of(origins))
+              .build();
+      assertEquals(2, rp.getOrigins().size());
+    }
   }
 
   @Test

--- a/webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyStartOperationSpec.scala
+++ b/webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyStartOperationSpec.scala
@@ -987,7 +987,7 @@ class RelyingPartyStartOperationSpec
           )
         )
         .preferredPubkeyParams(List(PublicKeyCredentialParameters.ES256).asJava)
-        .origins(Set.empty.asJava)
+        .origins(Set.empty[String].asJava)
       if (usernameRepository) {
         builder.usernameRepository(Helpers.UsernameRepository.withUsers(userId))
       }

--- a/webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyStartOperationSpec.scala
+++ b/webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyStartOperationSpec.scala
@@ -115,7 +115,7 @@ class RelyingPartyStartOperationSpec
         .identity(rpId)
         .credentialRepository(credRepo(credentials, userId))
         .preferredPubkeyParams(List(PublicKeyCredentialParameters.ES256).asJava)
-        .origins(Set.empty.asJava)
+        .origins(Set.empty[String].asJava)
       appId.foreach { appid => builder = builder.appId(appid) }
       attestationConveyancePreference.foreach { acp =>
         builder = builder.attestationConveyancePreference(acp)

